doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,222 @@
+Instead, we return `NULL`, so the caller can do something and doesn't have to deal with the internal bug.
+
+This overlapping of error and success values has especially positive effects, when we talk about boolean functions:
+
+```c
+// returns 1 if key is "foo", 0 if key is not "foo", and -1 if key == NULL
+int keyIsFooInt (Key * key);
+
+// returns true if key is "foo", false otherwise (including if key == NULL)
+bool keyIsFooBool (Key * key);
+```
+
+The `keyIsFooInt` function is very annoying to use, but `keyIsFooBool` is much easier except for the rare case where you need to distinguish a `NULL` key:
+
+```c
+// "foo" and not NULL
+if (keyIsFooInt (key) == 1)
+if (key != NULL && keyIsFooBool (key))
+
+// not "foo" or NULL
+if (keyIsFooInt (key) != 1)
+if (key == NULL || !keyIsFooBool (key))
+
+// not "foo" and not NULL
+if (keyIsFooInt (key) == 0)
+if (key != NULL && !keyIsFooBool (key))
+
+// "foo" or NULL
+if (keyIsFooInt (key) != 0)
+if (key == NULL || keyIsFooBool (key))
+```
+
+Especially, if you don't expect the `-1` return value (because `int` is also often boolean in C) bugs arise:
+
+```c
+// Expected: key is "foo"
+// Actually: key is "foo" or key is NULL
+if (keyIsFooInt (key))
+// Equivalent: if (keyIsFooInt (key) != 0)
+
+// Expected: key is not "foo"
+// Actually: key is not "foo" and key is not NULL
+if (!keyIsFooInt (key))
+// Equivalent: if (keyIsFooInt (key) == 0)
+```
+
+The above cases assume very simple preconditions.
+But sometimes the precondition checks are non-trivial.
+Take for example searching for `Key` in a `KeySet` with an escaped name:
+
+```c
+Key * ksLookupByName (KeySet * ks, const char * escapedName)
+```
+
+Because we return a pointer, the only error option is `NULL`.
+But that could mean multiple things:
+
+- No such key found
+- Error: `ks == NULL`
+- Error: `escapedName == NULL`
+- Error: `escapedName` not a valid escaped name
+
+If the caller wants to know which of these was the case, they have to do additional checks.
+For the first three cases that's easy, that's just two simple `NULL`-checks.
+But detecting the last case is non-trivial.
+Validating escaped names means checking a lot of rules.
+
+However, there is once again a very simple pragmatic solution ("everything is fine").
+We simply provide a public API function that allows checking, whether a `const char *` is a valid escaped name.
+Then the caller can just check after the error occured:
+
+```c
+// assume we already know ks != NULL && escapedName != NULL
+Key * found = ksLookupByName (ks, escapedName);
+if (found == NULL)
+{
+  if (!elektraIsValidEscapedName (escapedName)) {
+    // invalid name
+  } else {
+    // key not found
+  }
+}
+```
+
+## Decision
+
+**Suggestion:** Go with "Pragmatic solution"
+
+## Rationale
+
+Under the current assumptions this is the obvious solution.
+Every other solution borders on over-engineering.
+
+## Implications
+
+- Some functions (e.g., `keyIsBelow`) should have simplified return values
```

Comment:
I don't think this is an implication of that decision. If we use boolean for functions returning a boolean is imho a separate decision. I am open to that, I agree that it is weird to introduce a third case just for passing NULL.